daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/LawlessTraversalsSpec.scala

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.daml.lf.data
+
+import ImmArray.ImmArraySeq
+
+import org.scalatest.{Inside, Matchers, WordSpec}
+import org.scalatest.prop.GeneratorDrivenPropertyChecks
+
+@SuppressWarnings(Array("org.wartremover.warts.Any"))
+class LawlessTraversalsSpec
+    extends WordSpec
+    with Matchers
+    with Inside
+    with GeneratorDrivenPropertyChecks {
+  import LawlessTraversals._
+
+  "traverseEitherStrictly" should {
+    "satisfy identity, elementwise" in forAll { xs: Seq[Int] =>
+      xs traverseEitherStrictly (Right(_)) should ===(Right(xs))
+    }
+
+    "preserve class if the implementation bothered to set it up" in {
+      val classySeqs = Seq[Seq[Int]](List(1), Vector(2), ImmArraySeq(3))
+      // we need to use patmat, not == or shouldBe, because patmat is stricter
+      inside(classySeqs map (_ traverseEitherStrictly (Right(_)))) {
+        case Seq(_, Right(List(_)), _) => fail("lists are not vectors")
+        case Seq(Right(List(1)), Right(Vector(2)), Right(ImmArraySeq(3))) =>
```

Comment:
Noob Scala question, how does this work? The pattern does not seem exhaustive, and what does `case x =>` (without any value on the right hand side of `=>`) produce?

Reply:
> The pattern does not seem exhaustive

Pattern lists (`{ case pat => expr; ...}`) get two different treatments in the two kinds of lambda context. If the expected type is `T => R`, they get η-expanded like `x => x match {...}`, and normal exhaustiveness checking applies.

However, if the expected type is `T PartialFunction R`, exhaustiveness checking _does not_ apply, and the result gets a generated `isDefinedAt` (and compatible `applyOrElse` semantics) that returns true iff some `case` matching the argument is present. So the idiomatic way to write `catMaybes[A]: List[Option[A]] => List[A]` is `_ collect { case Some(x) => x }`.

`inside` takes a `PartialFunction`; as with `collect`, the inexhaustiveness is deliberate.

> what does `case x =>` (without any value on the right hand side of `=>`) produce?

`()`. This is usual for scalatest when you have nothing further to assert, see [`matchPattern` example](http://doc.scalatest.org/3.0.0/org/scalatest/Matchers.html#matchingAPattern).